### ISSUE TITLE
Shadow casting spot light flickering

### ISSUE THREAD

**Issue by Duion** [GarageGames/Torque3D#1871]
After updating my project to 3.10 ( and 3.9 ) I noticed that my night level where I had spotlights that cast shadows no longer work properly, they were flickering all the time and most of the time one or all of them were not working properly, here is a video showing it:
http://duion.com/files/temp/3.10-spotlight_flicker_bug.mp4

After I tested with a stock T3D release build I noticed the issue was not there, so I tried out different options.
I loaded the stock T3D engine into my project and the problem was still there.
Later I found out that this was caused by the sun/moon casting shadows, so when I disabled moon shadows it worked fine as before.
Then I tested stock T3D with sun/moon shadow and with spotlights and it still works, so the flaw had to be in my game.
Later I found out that it was the postFX setting SSAO that caused this, which was not enabled by default in stock T3D, so when I disabled SSAO it worked without flickering and with sun/moon light in my game project as well.
So any ideas why a level with SSAO + (Spotlight+castShadows) + (sun+castShadows) causes this?

**Comment by RichardsGameStudio** [comment]
Having the same problem with 3.9 but I think this is caused by reflections from water objects (lights get culled and not culled during subsequent reflection passes), and only at certain viewing angles. When switching of reflectivity of water object, the flickering does not occur. I made an console variable, which let me enable and disable the culling of light objects during reflection passes, although disabling it comes with a drop in fps (as can be expected).

**Comment by Duion** [comment]
Ah so it is the water, which makes sense, since I tested also in the default template desert level and this did not happen there.

**Comment by RichardsGameStudio** [comment]
Yes, as far as I can see it happens in scene/scenemanager.cpp _renderScene method during the reflections pass. Around line 429
you find this line where it all goes wrong.
`// Cull the list.`
`U32 numRenderObjects = state->getCullingState().cullObjects(`
`        mBatchQueryList.address(),`
`        mBatchQueryList.size(),`
`        !state->isDiffusePass() ? SceneCullingState::CullEditorOverrides : 0, // Keep forced editor stuff out of non-diffuse passes.`
`);`

**Comment by Azaezel** [comment]
unable to reproduce via https://gist.github.com/Azaezel/29d21166448fff755e1eb06d311337a0 mind modifying to a suitable testcase?

**Comment by RichardsGameStudio** [comment]
@Azaezel :
`!(state->isDiffusePass()` || state->isReflectPass()) ? SceneCullingState::CullEditorOverrides : 0
would be the alt to try if that is in fact the spot` does not solve the problem (at least for me).

My workaround: define a console variable which can be set/cleared using a trigger object. I use this trigger in cases where the flickering is annoying.
This console var I use as an add to the cullObjects method where it defaults to false.
Then in the cullObjects method I added:

     else if( !( cullOptions & DontCullRenderDisabled ) &&
               !object->isRenderEnabled() )
      {
         isCulled = true;
      }

	  // ME Add reflect disable culling of lights
	  else if (isReflect && object->getTypeMask()&LightObjectType)// ME Fix flickering lights
	  {
		  isCulled = false;
	  }

      // Global bounds objects are never culled.  Note that this means
      // that if these objects are to respect zoning, they need to manually
      // trigger the respective culling checks for whatever they want to
      // batch.

      else if( object->isGlobalBounds() )
         isCulled = false;


This works but it will drop fps when lights are not culled.

I will also look to see if I can make a testcase, but maybe @Duion can provide something quicker. If it does not occur in empty mission than perhaps it occurs when there is a certain number of objects in the mission.
